modules/plugin/shapefile/src/test/java/org/geotools/data/shapefile/DbaseFileTest.java

```diff
@@ -186,31 +188,79 @@ public void testEmptyFields() throws Exception {
+        assertEquals(sizeInBytes, formattedString.getBytes().length);
+        assertEquals("12345678", formattedString);
 
-        assertEquals("          ".getBytes().length, formattedString.getBytes().length);
+        // larger strings get trucated
+        formattedString = formatter.getFieldString(sizeInBytes, "12345678910");
+        assertEquals(sizeInBytes, formattedString.getBytes().length);
+        assertEquals("12345678", formattedString);
     }
 
     @Test
     public void testUTF8Chars() throws Exception {
-        // looks like "A, B, C" in an editor, but the B and C are actually weird UTF-8 entities
         DbaseFileWriter.FieldFormatter formatter =
                 new DbaseFileWriter.FieldFormatter(
                         StandardCharsets.UTF_8, TimeZone.getDefault(), false);
-        String test = "A, \u0412, \u0421";
+
+        int sizeInBytes = 4;
+
+        // a short string will be padded
+        String formattedString = formatter.getFieldString(sizeInBytes, "\u0412");
+        assertEquals(sizeInBytes, formattedString.getBytes().length);
+        assertEquals("\u0412  ", formattedString);
+
+        // a string of size btyes need no padding
+        formattedString = formatter.getFieldString(sizeInBytes, "\u0412\u0412");
+        assertEquals(sizeInBytes, formattedString.getBytes().length);
+        assertEquals("\u0412\u0412", formattedString);
+
+        // large strings get trucated
+        formattedString = formatter.getFieldString(sizeInBytes, "\u0412\u0412\u0412");
+        assertEquals(sizeInBytes, formattedString.getBytes().length);
+        assertEquals("\u0412\u0412", formattedString);
+
+        // if a multi-byte character is the last to be removed then padding may be required to
+        formattedString = formatter.getFieldString(sizeInBytes, "\u0412A\u0412\u0412");
+        assertEquals(sizeInBytes, formattedString.getBytes().length);
+        assertEquals("\u0412A ", formattedString);
+    }
+
+    @Test(timeout = 10000)
+    public void testVeryLongStrings() throws Exception {
+        // formatter.setFieldString will truncate input to the desired size. But it should do this
+        // in a reasonably performant manner.
+        DbaseFileWriter.FieldFormatter formatter =
+                new DbaseFileWriter.FieldFormatter(
+                        StandardCharsets.UTF_8, TimeZone.getDefault(), false);
+
+        // build up a very large input string. The test string is also formed so that the 8th char
+        // is also multibyte
+        String test =
+                IntStream.range(0, 100000)
+                        .mapToObj(i -> "\u0412A cat\u0412jumped over the dog")
+                        .collect(Collectors.joining(","));
+
         String formattedString = formatter.getFieldString(8, test);
-        assertEquals("        ".getBytes().length, formattedString.getBytes().length);
+        // for this test we are more concerned with getting the answer in a reasonable time, rather
```

Comment:
But you are testing the actual answer too, so the comment here is confusing, suggest removing it.

Reply:
will remove